File: reticle-agent/android/src/main/kotlin/dev/reticle/agent/LottieBridge.kt
package dev.reticle.agent

import android.graphics.Paint
import android.view.View
import android.widget.ImageView
import dev.reticle.core.InteractionRegion
import dev.reticle.core.Rect
import dev.reticle.core.RegionSource

/**
 * Looks inside a Lottie animation.
 *
 * A Lottie renders its whole UI — text and shapes — into one opaque canvas, so
 * the view tree sees a single `LottieAnimationView` node. When an app bakes a
 * dialog's title / message / buttons into the animation, none of that is a real
 * view, an accessibility node, or a DOM element, so nothing downstream can act
 * on it. This bridge recovers those elements from the *parsed composition* Lottie
 * already holds in memory: it enumerates the named layers, reads text layers'
 * strings, and maps each layer's transform through the composition→view scale to
 * a screen rect — surfaced as [RegionSource.lottie] sub-regions so the existing
 * region pipeline (`ui regions`, `act tap --region`) can target them.
 *
 * Pure reflection: the agent must not link Lottie (a target app may not use it).
 * Every step is guarded — an absent class, a renamed field, or an unexpected
 * shape yields no regions rather than a crash. Geometry comes from the same model
 * Lottie draws from; the "this layer is a button" inference is a hint (the label
 * is the layer's own text, or its author-given name).
 */
object LottieBridge {

    private const val VIEW_CLASS = "com.airbnb.lottie.LottieAnimationView"

    fun regionsFor(view: View, screenFrame: Rect?): List<InteractionRegion> {
        if (screenFrame == null) return emptyList()
        if (!isLottieView(view)) return emptyList()
        return try {
            val composition = invoke(view, "getComposition") ?: return emptyList()
            val bounds = invoke(composition, "getBounds") ?: return emptyList()
            val compW = (invoke(bounds, "width") as? Int)?.toDouble() ?: return emptyList()
            val compH = (invoke(bounds, "height") as? Int)?.toDouble() ?: return emptyList()
            if (compW <= 0 || compH <= 0) return emptyList()

            @Suppress("UNCHECKED_CAST")
            val layers = invoke(composition, "getLayers") as? List<Any?> ?: return emptyList()
            val map = ViewMap.of(view, screenFrame, compW, compH)

            layers.mapNotNull { layer -> layer?.let { regionForLayer(it, map) } }
        } catch (_: Throwable) {
            emptyList()
        }
    }

    private fun isLottieView(view: View): Boolean {
        var c: Class<*>? = view.javaClass
        while (c != null) {
            if (c.name == VIEW_CLASS) return true
            c = c.superclass
        }
        return false
    }

    /** One text layer -> one region. Non-text layers are skipped (no label). */
    private fun regionForLayer(layer: Any, map: ViewMap): InteractionRegion? {
        val type = invoke(layer, "getLayerType")?.let { (it as? Enum<*>)?.name }
        if (type != "TEXT") return null

        val textFrame = accessibleInvoke(layer, "getText") ?: return null
        @Suppress("UNCHECKED_CAST")
        val keyframes = invoke(textFrame, "getKeyframes") as? List<Any?> ?: return null
        val doc = keyframes.firstNotNullOfOrNull { it?.let { kf -> field(kf, "startValue") } } ?: return null
        val text = (field(doc, "text") as? String)?.takeIf { it.isNotBlank() } ?: return null
        val size = (field(doc, "size") as? Float)?.toDouble() ?: 12.0
        val justify = (field(doc, "justification") as? Enum<*>)?.name ?: "LEFT_JUSTIFY"

        val transform = accessibleInvoke(layer, "getTransform") ?: return null
        val pos = firstPoint(invoke(transform, "getPosition")) ?: return null
        val anchor = firstPoint(invoke(transform, "getAnchorPoint")) ?: (0.0 to 0.0)
        val cx = pos.first - anchor.first
        val cy = pos.second - anchor.second

        // Measure the real glyph extent instead of estimating from char count —
        // accurate for proportional fonts and CJK/Latin/emoji runs.
        val paint = Paint().apply { textSize = size.toFloat() }
        val glyphW = paint.measureText(text).toDouble()
        val fm = paint.fontMetrics
        val glyphH = (fm.descent - fm.ascent).toDouble()

        // The authored text box (`sz`/`ps` in the Lottie), when present, gives the
        // exact anchor + layout region the glyphs are placed within; fall back to
        // the layer origin otherwise.
        val boxLeft = cx + (pointField(field(doc, "boxPosition"), "x") ?: 0.0)
        val boxTop = cy + (pointField(field(doc, "boxPosition"), "y") ?: 0.0)
        val boxW = pointField(field(doc, "boxSize"), "x")?.takeIf { it > 0 }
        val boxH = pointField(field(doc, "boxSize"), "y")?.takeIf { it > 0 }

        // Place the measured glyph rect by justification within the box (or at the
        // origin when unboxed), so the rect hugs the text and its center is a
        // reliable tap point.
        val anchorX = when (justify) {
            "CENTER" -> boxW?.let { boxLeft + it / 2 } ?: cx
            "RIGHT_JUSTIFY" -> boxW?.let { boxLeft + it } ?: cx
            else -> if (boxW != null) boxLeft else cx
        }
        val left = when (justify) {
            "CENTER" -> anchorX - glyphW / 2
            "RIGHT_JUSTIFY" -> anchorX - glyphW
            else -> anchorX
        }
        val centerY = boxH?.let { boxTop + it / 2 } ?: cy
        val top = centerY - glyphH / 2

        return InteractionRegion(
            source = RegionSource.lottie,
            label = text,
            rects = listOf(map.rect(left, top, glyphW, glyphH)),
        )
    }

    /** A public `PointF` field's component as a Double, or null. */
    private fun pointField(point: Any?, axis: String): Double? =
        point?.let { (field(it, axis) as? Float)?.toDouble() }

    /** Composition-space -> screen-space mapping for the view's scaleType. */
    private class ViewMap(
        val originX: Double, val originY: Double,
        val offX: Double, val offY: Double,
        val sx: Double, val sy: Double,
    ) {
        fun rect(cxLeft: Double, cyTop: Double, w: Double, h: Double): Rect = Rect(
            x = originX + offX + cxLeft * sx,
            y = originY + offY + cyTop * sy,
            width = w * sx,
            height = h * sy,
        )

        companion object {
            fun of(view: View, frame: Rect, compW: Double, compH: Double): ViewMap {
                val fitXy = (view as? ImageView)?.scaleType == ImageView.ScaleType.FIT_XY
                return if (fitXy) {
                    ViewMap(frame.x, frame.y, 0.0, 0.0, frame.width / compW, frame.height / compH)
                } else {
                    // FIT_CENTER / CENTER_INSIDE — the LottieAnimationView default:
                    // uniform min-scale, centered.
                    val s = minOf(frame.width / compW, frame.height / compH)
                    ViewMap(
                        frame.x, frame.y,
                        (frame.width - compW * s) / 2, (frame.height - compH * s) / 2,
                        s, s,
                    )
                }
            }
        }
    }

    /** First keyframe's start value of an AnimatablePathValue, as (x, y). */
    private fun firstPoint(pathValue: Any?): Pair<Double, Double>? {
        pathValue ?: return null
        @Suppress("UNCHECKED_CAST")
        val keyframes = (invoke(pathValue, "getKeyframes") as? List<Any?>) ?: return null
        val point = keyframes.firstNotNullOfOrNull { it?.let { kf -> field(kf, "startValue") } } ?: return null
        val x = (field(point, "x") as? Float)?.toDouble() ?: return null
        val y = (field(point, "y") as? Float)?.toDouble() ?: return null
        return x to y
    }

    // --- reflection helpers (all null-safe) ---

    private fun invoke(target: Any, method: String): Any? = try {
        target.javaClass.getMethod(method).invoke(target)
    } catch (_: Throwable) {
        null
    }

    /** Invoke a (possibly package-private) no-arg method, walking up the class. */
    private fun accessibleInvoke(target: Any, method: String): Any? {
        var c: Class<*>? = target.javaClass
        while (c != null) {
            try {
                val m = c.getDeclaredMethod(method)
                m.isAccessible = true
                return m.invoke(target)
            } catch (_: NoSuchMethodException) {
                c = c.superclass
            } catch (_: Throwable) {
                return null
            }
        }
        return null
    }

    private fun field(target: Any, name: String): Any? {
        var c: Class<*>? = target.javaClass
        while (c != null) {
            try {
                val f = c.getDeclaredField(name)
                f.isAccessible = true
                return f.get(target)
            } catch (_: NoSuchFieldException) {
                c = c.superclass
            } catch (_: Throwable) {
                return null
            }
        }
        return null
    }
}

File: reticle-agent/ios/Sources/ReticleKit/LottieBridge.swift
import UIKit
import ReticleProtocol

/// Looks inside a Lottie animation — the iOS counterpart of the Android
/// `LottieBridge`.
///
/// A Lottie renders its whole UI into one layer, so the view tree sees a single
/// `LottieAnimationView`. When an app bakes a dialog's title / message / buttons
/// into the animation, none of it is a real view, so nothing downstream can act
/// on it. This recovers those elements from the *parsed model* Lottie holds in
/// memory: it enumerates the text layers, reads their strings, and maps each
/// layer's transform through the composition→view scale to a screen rect —
/// surfaced as `.lottie` sub-regions for the existing region pipeline.
///
/// Pure `Mirror` reflection: the agent must not link Lottie. iOS has no public
/// introspection API and Swift can't invoke a non-`@objc` method by name, so
/// this reads stored properties only. Every step is optional-guarded — a renamed
/// property or an unexpected shape yields no regions rather than a crash. The
/// geometry comes from the same model Lottie draws from; the "this layer is a
/// button" inference is a hint (the label is the layer's own text).
///
/// `@MainActor` because it reads UIView state (`contentMode`) to build the
/// composition→screen map, like every other capture path here. Under Swift 6.1's
/// UIKit annotations that read from a nonisolated context is an error; a newer
/// toolchain accepts it, which is how this compiled locally while CI (Xcode 16.4)
/// rejected it. The whole capture is main-thread work anyway — `SnapshotCapture` is
/// already `@MainActor` and is the only caller.
@MainActor
enum LottieBridge {

    static func regions(for view: UIView, screenFrame: Rect?) -> [InteractionRegion] {
        guard let screenFrame else { return [] }
        guard NSStringFromClass(type(of: view)).contains("LottieAnimationView") else { return [] }
        guard
            let layer = mchild(view, "lottieAnimationLayer"),
            let animation = mchild(layer, "animation"),
            let compW = mchild(animation, "width") as? Double,
            let compH = mchild(animation, "height") as? Double,
            compW > 0, compH > 0,
            let layersAny = mchild(animation, "layers")
        else { return [] }

        let map = ViewMap(view: view, frame: screenFrame, compW: compW, compH: compH)
        return melements(layersAny).compactMap { region(for: $0, map: map) }
    }

    /// A text layer (identified by the presence of a `text` keyframe group) →
    /// one region. Non-text layers carry no readable label and are skipped.
    private static func region(for layer: Any, map: ViewMap) -> InteractionRegion? {
        guard
            let textGroup = mchild(layer, "text"),
            let kfs = mchild(textGroup, "keyframes"),
            let firstKf = melements(kfs).first,
            let doc = mchild(firstKf, "value"),
            let text = mchild(doc, "text") as? String, !text.isEmpty
        else { return nil }

        let fontSize = (mchild(doc, "fontSize") as? Double) ?? 12
        let transform = mchild(layer, "transform")
        let pos = transform.flatMap { firstXY(mchild($0, "position")) } ?? (0, 0)
        let anchor = transform.flatMap { firstXY(mchild($0, "anchorPoint")) } ?? (0, 0)
        let originX = pos.0 - anchor.0
        let originY = pos.1 - anchor.1

        // Prefer the authored text box; fall back to an estimate from the string.
        let framePos = xy(mchild(doc, "textFramePosition"))
        let frameSize = xy(mchild(doc, "textFrameSize"))
        let boxW: Double
        let boxH: Double
        let leftComp: Double
        let topComp: Double
        if let framePos, let frameSize, frameSize.0 > 0, frameSize.1 > 0 {
            boxW = frameSize.0
            boxH = frameSize.1
            leftComp = originX + framePos.0
            topComp = originY + framePos.1
        } else {
            // No authored box: estimate from the string, centered on the layer
            // position (the common center-justified authoring).
            boxW = Double(text.count) * fontSize * 0.6
            boxH = fontSize * 1.2
            leftComp = originX - boxW / 2
            topComp = originY - boxH / 2
        }

        return InteractionRegion(
            source: .lottie,
            label: text,
            rects: [map.rect(leftComp, topComp, boxW, boxH)]
        )
    }

    /// Composition-space → screen-space for the view's contentMode.
    private struct ViewMap {
        let ox: Double, oy: Double, sx: Double, sy: Double
        init(view: UIView, frame: Rect, compW: Double, compH: Double) {
            let vw = frame.width, vh = frame.height
            switch view.contentMode {
            case .scaleToFill:
                sx = vw / compW; sy = vh / compH
                ox = frame.x; oy = frame.y
            case .scaleAspectFill:
                let s = max(vw / compW, vh / compH)
                sx = s; sy = s
                ox = frame.x + (vw - compW * s) / 2
                oy = frame.y + (vh - compH * s) / 2
            default: // .scaleAspectFit — the LottieAnimationView default
                let s = min(vw / compW, vh / compH)
                sx = s; sy = s
                ox = frame.x + (vw - compW * s) / 2
                oy = frame.y + (vh - compH * s) / 2
            }
        }
        func rect(_ l: Double, _ t: Double, _ w: Double, _ h: Double) -> Rect {
            Rect(x: ox + l * sx, y: oy + t * sy, width: w * sx, height: h * sy)
        }
    }

    // MARK: - Mirror helpers (read stored properties only; never throw)

    /// A stored child by label, unwrapping one level of Optional. Walks the
    /// superclass chain — `transform`/`name` live on the `LayerModel` base, and
    /// `Mirror` only lists a class's own stored properties by default. Nil if
    /// absent.
    private static func mchild(_ any: Any, _ label: String) -> Any? {
        var mirror: Mirror? = Mirror(reflecting: any)
        while let m = mirror {
            for c in m.children where c.label == label {
                return unwrap(c.value)
            }
            mirror = m.superclassMirror
        }
        return nil
    }

    private static func unwrap(_ any: Any) -> Any? {
        let m = Mirror(reflecting: any)
        guard m.displayStyle == .optional else { return any }
        guard let inner = m.children.first?.value else { return nil }
        return unwrap(inner)
    }

    private static func melements(_ any: Any) -> [Any] {
        Mirror(reflecting: any).children.map { $0.value }
    }

    /// (x, y) of the first keyframe value in a KeyframeGroup<LottieVector3D>.
    private static func firstXY(_ group: Any?) -> (Double, Double)? {
        guard let group, let kfs = mchild(group, "keyframes"),
              let first = melements(kfs).first, let val = mchild(first, "value") else { return nil }
        return xy(val)
    }

    /// (x, y) of a LottieVector3D.
    private static func xy(_ vec: Any?) -> (Double, Double)? {
        guard let vec, let x = mchild(vec, "x") as? Double, let y = mchild(vec, "y") as? Double else { return nil }
        return (x, y)
    }
}
